KAFKA-8443 Broker support for fetch from followers

mumrah opened the pull request:
Follow on to #6731, this PR adds broker-side support for [KIP-392](https://cwiki.apache.org/confluence/display/KAFKA/KIP-392%3A+Allow+consumers+to+fetch+from+closest+replica) (fetch from followers). 

Changes:
* All brokers will handle FetchRequest regardless of leadership
* Leaders can compute a preferred replica to return to the client
* New ReplicaSelector interface for determining the preferred replica
* Incremental fetches will include partitions with no records if the preferred replica has been computed
* Adds new JMX to expose the current preferred read replica of a partition in the consumer

Two new conditions were added for completing a delayed fetch. They both relate to communicating the high watermark to followers without waiting for a timeout:
* For regular fetches, if the high watermark changes within a single fetch request 
* For incremental fetch sessions, if the follower's high watermark is lower than the leader

A new JMX attribute `preferred-read-replica` was added to the `kafka.consumer:type=consumer-fetch-manager-metrics,client-id=some-consumer,topic=my-topic,partition=0` object. This was added to support the new system test which verifies that the fetch from follower behavior works end-to-end. This attribute could also be useful in the future when debugging problems with the consumer.

mumrah committed:
Copying work over from prototype PR

mumrah committed:
Don't check replica ID when handling fetch requests

hachikuji commented on clients/src/main/java/org/apache/kafka/common/replica/ReplicaSelector.java:
```diff
@@ -0,0 +1,85 @@
+     */
+    Optional<ReplicaInfo> select(TopicPartition topicPartition,
+                                 ClientMetadata clientMetadata,
+                                 Set<ReplicaInfo> replicaInfos);
```
Would it make sense to wrap this in a `PartitionInfo` object? That will be easier to evolve if we find the need for new information which is at the partition level instead of the replica level. For example, we could have a separate accessor for the leader and take `isLeader` out of `ReplicaInfo`.

mumrah replied:
Do you mean the return type or the 3rd argument? The problem with taking PartitionInfo as an argument to `select` is it doesn't have some of the info we want like last caught up time and LEO.

As a return type, that would work fine I think since we mainly care about the `Node` for the replica.

hachikuji replied:
I was thinking more conceptually, not necessarily reusing `PartitionInfo`. We should have an object which represents the partition state.

mumrah replied:
Ah, I see what you mean, yea that seems like a good idea.

mumrah committed:
Add PartitionInfo to ReplicaSelector interface

And some tests

mumrah committed:
Rename some classes

mumrah committed:
backing out a file

mumrah committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-8443

mumrah committed:
Fix scala 2.11 compile error

hachikuji commented on clients/src/main/java/org/apache/kafka/common/replica/ReplicaSelector.java:
```diff
@@ -0,0 +1,99 @@
+    /**
+     * Holder for all the client metadata required to determine a preferred replica.
+     */
+    class ClientMetadata {
```
Does it makes sense to stick an interface behind this also?

mumrah replied:
Sure, done.

mumrah committed:
Update FetchSession logic

mumrah committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-8443

mumrah committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-8443

mumrah commented:
retest this please

mumrah committed:
Expose partition read replica over JMX and add system test

mumrah committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-8443

mumrah commented:
retest this please

mumrah commented on core/src/main/scala/kafka/server/MetadataCache.scala:
```diff
@@ -195,6 +197,23 @@ class MetadataCache(brokerId: Int) extends Logging {
+    snapshot.partitionStates.get(topic).flatMap(_.get(partitionId)).map(partitionInfo => {
+      val replicaIds = partitionInfo.basePartitionState.replicas
+      replicaIds.asScala
+      //JavaConverters.asScalaBuffer(replicaIds)
```
Clean this up

mumrah commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1000,6 +1016,46 @@ class ReplicaManager(val config: KafkaConfig,
+        val replicaEndpoints = metadataCache.getPartitionReplicaEndpoints(tp.topic(), tp.partition(), new ListenerName(clientMetadata.listenerName))
+        val replicaInfoSet: Set[ReplicaView] = partition.allReplicas.map(
+          replica => SomeReplicaView(
+            isLeader = partition.leaderReplicaIdOpt.exists(leaderId => leaderId.equals(replica.brokerId)),
```
@hachikuji is this a reasonable way to test if the local replica is the leader?

hachikuji replied:
Seems reasonable, but as suggested above, I wonder if we can get rid of this field.

mumrah replied:
I changed it to a default field based on the set of ReplicaView-s. If we really don't want it as a member of the interface I can move it to a static. It is convenient to quickly get the leader optional from the PartitionView

mumrah replied:
Moved to a static method on ReplicaSelector

hachikuji commented on clients/src/main/java/org/apache/kafka/common/replica/MostCaughtUpReplicaSelector.java:
```diff
@@ -0,0 +1,34 @@
+import java.util.Comparator;
+import java.util.Optional;
+
+public class MostCaughtUpReplicaSelector implements ReplicaSelector {
```
Do you think this selector has any benefit outside of breaking ties in `RackAwareReplicaSelector`?

mumrah replied:
No not really. I'll take it out (I was thinking about removing it anyways).

hachikuji commented on clients/src/main/java/org/apache/kafka/common/replica/ReplicaSelector.java:
```diff
@@ -0,0 +1,102 @@
+
+        Node endpoint();
+
+        long logOffset();
```
I think this is meant to be the end offset. Maybe we can name it `logEndOffset` or just `endOffset`?

mumrah replied:
Good catch, will fix

hachikuji commented on clients/src/main/java/org/apache/kafka/common/replica/RackAwareReplicaSelector.java:
```diff
@@ -0,0 +1,64 @@
+                            return partitionView.leader();
+                        }
+                    };
+                    return tieBreaker.select(topicPartition, clientMetadata, sameRackPartition);
```
This could return a replica which is arbitrarily far behind the leader. I think what we suggested in the KIP was that we would limit selection to the ISR. We also discussed taking the fetch offset into account. It may be totally fine to return an out-of-sync replica as long as it has the requested offset. I rather liked this latter idea because it ensured we wouldn't get into any redirection loops where the consumer is switching back and forth between the leader and follower. Potentially the broker could do this filtering before passing the replicas to the selector which would avoid the extra api noise.

mumrah replied:
So, we just need to filter the set of replicas before passing to the selector? Seems straightforward.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/replica/ReplicaSelector.java:
```diff
@@ -0,0 +1,102 @@
+
+        long logOffset();
+
+        long lastCaughtUpTimeMs();
```
Conceivably a replica may never have been caught up. I guess our options are either to make this an Optional or use the -1 sentinel.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/replica/ReplicaSelector.java:
```diff
@@ -0,0 +1,102 @@
+     * View of a replica used by {@link ReplicaSelector} to determine a preferred replica.
+     */
+    interface ReplicaView {
+        boolean isLeader();
```
Can we get rid of this since we expose the leader directly from `PartitionView`?

hachikuji commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1000,6 +1016,46 @@ class ReplicaManager(val config: KafkaConfig,
+          ))
+        val partitionInfo = SomePartitionView(
+          replicas = replicaInfoSet.asJava,
+          leader = CoreUtils.asJavaOptional(replicaInfoSet.find(info => info.isLeader))
```
I think you can now use `asJava` with `import scala.compat.java8.OptionConverters._`

hachikuji commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -829,16 +838,16 @@ class ReplicaManager(val config: KafkaConfig,
       val result = readFromLocalLog(
         replicaId = replicaId,
-        fetchOnlyFromLeader = fetchOnlyFromLeader,
+        fetchOnlyFromLeader = false, // TODO remove this argument
```
Should we continue to enforce `fetchOnlyFromLeader` if the fetch is from a follower?

mumrah replied:
Yes, we probably should.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetcherMetricsRegistry.java:
```diff
@@ -139,7 +140,9 @@ public FetcherMetricsRegistry(Set<String> tags, String metricGrpPrefix) {
                 "The average lead of the partition", partitionTags);
-
+        this.partitionPreferredReadReplica = new MetricNameTemplate(
+                "preferred-read-replica", "consumer-fetch-manager-metrics",
```
We'll need to update the KIP at some point with this and the other API changes.

mumrah replied:
Done

hachikuji commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -926,6 +937,10 @@ class ReplicaManager(val config: KafkaConfig,
 
+
+        // If we are the leader, determine the preferred read-replica
+        val preferredReadReplica = findPreferredReadReplica(tp, clientMetadata, replicaId)
```
We are choosing the preferred read replica after reading from the log. I wonder if we should do it first instead. This is related to my other comment, but I think the logic can be something like this:

1. Find all replicas which have data at the fetch offset.
2. If this set is more than just the leader, the select the preferred read replica.
3. If the preferred read replica is not the leader, return empty and skip the read.
4. Otherwise do the read as usual.

hachikuji replied:
One more thing, it seems we are selecting the preferred read replica even if we are not the leader? Should we skip this if we are the follower?

mumrah replied:
The logic for checking if we're the leader is inside `findPreferredReadReplica`. I'll update the read logic as suggested

hachikuji reviewed (commented):
Thanks, looks pretty clean. Left a few comments.

mumrah committed:
Feedback from PR on ReplicaSelector interface

mumrah committed:
Feedback from PR for ReplicaManager

mumrah committed:
More PR feedback, fix the tests

mumrah committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-8443

mumrah committed:
Fix logic error in ReplicaManager

hachikuji commented on clients/src/main/java/org/apache/kafka/common/replica/ReplicaSelector.java:
```diff
@@ -0,0 +1,145 @@
+         * Helper method to find the leader among the replicas for this partition
+         */
+        default Optional<ReplicaView> findLeader() {
+            return replicas().stream().filter(ReplicaView::isLeader).findFirst();
```
It's a bit subjective, but I actually preferred having the leader state as a partition-level trait. It expresses cleanly the fact that there can be only one leader.

mumrah replied:
Ok, think a helper method would be more suitable then (in order to keep this out of the contract)?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -1706,8 +1708,20 @@ private void maybeUpdateAssignment(SubscriptionState subscription) {
                 }
+
+                for (TopicPartition tp : newAssignedPartitions) {
+                    if (!this.assignedPartitions.contains(tp)) {
```
Not a big deal, but could potentially skip this check since we are checking for metric existence below anyway.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -1759,6 +1773,12 @@ private static String partitionLeadMetricName(TopicPartition tp) {
         }
 
+        private MetricName partitionPreferredReadReplicaMetricName(TopicPartition tp) {
+            Map<String, String> metricTags = new HashMap<>(2);
```
nit: We have this logic in a couple other places. Maybe sufficient cause to turn it into a method?

hachikuji commented on clients/src/main/java/org/apache/kafka/common/replica/ReplicaSelector.java:
```diff
@@ -0,0 +1,145 @@
+        /**
+         * The number of milliseconds (if any) since the last time this replica was caught up to the high watermark.
+         */
+        Optional<Long> lastCaughtUpTimeMs();
```
Perhaps we can document what this returns for the leader?

mumrah replied:
Hmm, what would that be exactly? The fetch time? Or would it remain at zero for the leader replica?

hachikuji replied:
Yes, that is what I was wondering 😉 . Possibly this should be a delta since the last caught up time or something. Then for the leader, it would always be 0.

mumrah replied:
Looks like this value only gets updated for follower replicas, so the leader will have the initial value of `0L`. This maps to `Optional.empty()` for a ReplicaView. Sound reasonable?

hachikuji replied:
Did you not like my suggestion to treat this as a delta since the last time we were caught up? In that case, the value is just always 0 for a leader.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/replica/ReplicaSelector.java:
```diff
@@ -0,0 +1,145 @@
+                .thenComparing(replicaInfo -> replicaInfo.endpoint().id());
+        }
+    }
+}
```
nit: newline

mumrah commented:
Retest this please

hachikuji commented on core/src/main/scala/kafka/server/DelayedFetch.scala:
```diff
@@ -114,6 +122,15 @@ class DelayedFetch(delayMs: Long,
               }
             }
+
+            // Case G check if HW has changed
```
I think this is handling the case when the high watermark changes while we are awaiting a fetch. It doesn't handle the case when the high watermark has been bumped between fetches. I had thought that we may be able to keep track of the last observed high watermark of a follower in the fetch session. When the fetch session is first initialized, we could always respond immediately to be sure that the follower has an updated value.

I did feel a little uneasy about the performance implications of this change since it can only increase fetch overhead. The fetch session should make the impact not too bad, but it's hard to be sure. For high throughput topics, we should always have data to return anyway, so the difference is probably negligible. The worst case is low volume topics and we would see roughly twice as many fetches as before (one fetch to get the data, one fetch to get the updated high watermark). But if the volume is low, then the additional overhead probably doesn't matter.

mumrah replied:
Updated the DelayedFetch to include another case for this. We will try to lookup the high watermark for a follower in the fetch session (only incremental fetch). If the offset snapshot from the leader has a higher HW, we'll complete the fetch.

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -82,11 +84,18 @@ import org.apache.kafka.common.security.token.delegation.{DelegationToken, Token
 import scala.collection.mutable.ArrayBuffer
 import scala.util.{Failure, Success, Try}
 
+case class SomeClientMetadata(rackId: String,
```
I don't suppose there's somewhere else we can put this? `KafkaApis` is already pretty cluttered. Would it be reasonable to add a default implementation in the Java code. The only apparent downside is that we wouldn't be able to use pattern matching.

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -577,6 +586,13 @@ class KafkaApis(val requestChannel: RequestChannel,
       fetchRequest.isFromFollower)
 
+    val clientMetadata = SomeClientMetadata(
+      fetchRequest.rackId(),
```
nit: unneeded parenthesis

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -671,8 +689,15 @@ class KafkaApis(val requestChannel: RequestChannel,
         val lastStableOffset = data.lastStableOffset.getOrElse(FetchResponse.INVALID_LAST_STABLE_OFFSET)
-        partitions.put(tp, new FetchResponse.PartitionData(data.error, data.highWatermark, lastStableOffset,
-          data.logStartOffset, abortedTransactions, data.records))
+        if (data.preferredReadReplica.isDefined) {
```
Could we skip this check since we're mapping to an Optional anyway?

mumrah replied:
Good point

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -577,6 +586,13 @@ class KafkaApis(val requestChannel: RequestChannel,
       fetchRequest.toForget,
       fetchRequest.isFromFollower)
 
+    val clientMetadata = SomeClientMetadata(
```
One thing we probably have to be a little careful about is ensuring that the fetch API version supports preferred read replicas. Possibly we could check that here and then pass `Option[ClientMetadata]` into `fetchMessages`. There are probably other ways as well, but I think this would let us get rid of `ClientMetadata.NO_METADATA`.

mumrah replied:
Good suggestion 👍 done.

hachikuji commented on core/src/main/scala/kafka/server/MetadataCache.scala:
```diff
@@ -195,6 +197,22 @@ class MetadataCache(brokerId: Int) extends Logging {
     }
   }
 
+  def getPartitionReplicaEndpoints(topic: String, partitionId: Int, listenerName: ListenerName): Map[Int, Node] = {
```
nit: may as well let this take `TopicPartition` since that is what the caller has anyway

hachikuji commented on core/src/main/scala/kafka/server/MetadataCache.scala:
```diff
@@ -195,6 +197,22 @@ class MetadataCache(brokerId: Int) extends Logging {
+            case None =>
+              Node.noNode()
+          }}).toMap
+        .filter(pair => !pair._2.equals(Node.noNode()))
```
Node has an `isEmpty` method we may be able to use.

hachikuji commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -875,9 +885,9 @@ class ReplicaManager(val config: KafkaConfig,
         })
       }
-      val fetchMetadata = FetchMetadata(fetchMinBytes, fetchMaxBytes, hardMaxBytesLimit, fetchOnlyFromLeader,
+      val fetchMetadata = FetchMetadata(fetchMinBytes, fetchMaxBytes, hardMaxBytesLimit, false,
```
Should we revert this?

hachikuji commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -917,35 +928,53 @@ class ReplicaManager(val config: KafkaConfig,
+          // If the a preferred read-replica is set and is not this replica (the leader), skip the read
+          val localReplica: Replica = partition.localReplicaOrException
+          LogReadResult(info = FetchDataInfo(LogOffsetMetadata.UnknownOffsetMetadata, MemoryRecords.EMPTY),
+            highWatermark = localReplica.highWatermark,
```
Can we use `Partition.fetchOffsetSnapshot` to get these offsets? This ensures that we always go through the epoch check.

hachikuji commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1000,6 +1029,47 @@ class ReplicaManager(val config: KafkaConfig,
+            }
+          ))
+        val partitionInfo = SomePartitionView(replicas = replicaInfoSet.asJava)
+        Option.apply(replicaSelector.select(tp, clientMetadata, partitionInfo).orElse(null))
```
nit: you can drop `.apply`

hachikuji commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1000,6 +1029,47 @@ class ReplicaManager(val config: KafkaConfig,
+    if (partition.leaderReplicaIfLocal.isDefined) {
+      if (Request.isValidBrokerId(replicaId)) {
+        // Don't look up preferred for follower fetches via normal replication
+        Option.empty
```
nit: i think it's more common to use `None` directly

hachikuji reviewed (commented):
Thanks for the updates. Did another pass.

mumrah committed:
More feedback from PR. Moving stuff around

mumrah committed:
Fixup

mumrah committed:
Add missing case for fetch satisfaction

mumrah committed:
More feedback from PR

mumrah commented:
retest this please

mumrah commented:
retest this please

mumrah committed:
Fix fetch request test

mumrah committed:
Add a couple unit tests for new DelayedFetch behavior

mumrah committed:
Cleanup

jsancio commented on clients/src/main/java/org/apache/kafka/common/replica/RackAwareReplicaSelector.java:
```diff
@@ -0,0 +1,56 @@
+                                        PartitionView partitionView) {
+        if (clientMetadata.rackId() != null && !clientMetadata.rackId().isEmpty()) {
+            Set<ReplicaView> sameRackReplicas = partitionView.replicas().stream()
+                    .filter(replicaInfo -> clientMetadata.rackId().equalsIgnoreCase(replicaInfo.endpoint().rack()))
```
nit: Do we mean to ignore the case? If so I would like to understand the reason. Up to this point I was assuming that we handle broker.rack as an opaque value for which we simply perform string comparison. For example, here is some code that uses the rack property.

Rack aware assignment: https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/admin/AdminUtils.scala#L207-L213

jsancio replied:
Also, why is this in the `client` project? I was under the impression that we wanted to make selection decision on the server.

mumrah replied:
Thanks, that's a good catch. I'll preserve the case in the comparison. As for the `client` project, that's a good question. Since this is server impl code, it could also go in the core project, but then we'd need to rewrite it in Scala. @hachikuji wdyt?

hachikuji replied:
I think we don't want users depending on kafka-core. If we had a separate package for only server plugins, that might be another approach, but we have traditionally put common client and server code in kafka-clients. As for scala/java, we tend to prefer exposed interfaces to be written in Java since since that is easier for most users to work with. Using scala from java can be a bit of a bear, for example, but the opposite is usually straightforward.

jsancio commented on core/src/main/scala/kafka/server/DelayedFetch.scala:
```diff
@@ -59,9 +60,16 @@ class DelayedFetch(delayMs: Long,
+  /**
+    * Keep track of the first high watermark we encounter during this fetch for each partition
+    */
+  val seenHighWatermarks: mutable.Map[TopicPartition, Long] = mutable.Map()
```
In Scala you can `private[this] val seenHighWatermarks = mutable.Map.empty[TopicPartition, Long]`

jsancio reviewed (commented):
Thanks for the PR. Quick high-level review.

mumrah committed:
More feedback

hachikuji commented on core/src/main/scala/kafka/server/DelayedFetch.scala:
```diff
@@ -71,7 +79,8 @@ class DelayedFetch(delayMs: Long,
    * Case F: This broker is the leader, but the requested epoch is now fenced
-   *
+   * Case G: The high watermark on this broker has changed within a FetchSession, need to propagate to follower (KIP-392)
+   * Case H: The high watermark on this broker has changed during this request, need to propagate to followers (KIP-392)
```
Not sure I understand why this should be a separate case. Wouldn't it be subsumed by G?

mumrah replied:
I guess if this was a fetch from a normal consumer and not a follower, we wouldn't have the high watermark information for case G

mumrah replied:
Oh, but we only do this for followers anyways. Nevermind

hachikuji commented on clients/src/main/java/org/apache/kafka/common/replica/ReplicaSelector.java:
```diff
@@ -0,0 +1,56 @@
+    /**
+     * Helper method to find the leader among the replicas for this partition
+     */
+    static Optional<ReplicaView> findLeader(PartitionView partitionView) {
```
I guess I'm not having much luck with this argument 😉. My interest is both reducing the cost of `LeaderReplicaSelector` and representing the partition state in a simple way. The cost argument is a bit nebulous, but fetching is on the hot path and I'm inclined to reduce unnecessary work. From an API clarity point of view, I still think it's simpler to take `isLeader` out of replica and just have a single field at the partition level for the leader. Is there something I'm missing?

mumrah replied:
Sorry, I was being dense. I understand what you're saying (finally). I'll move the leader field/method to partition 👍

hachikuji commented on core/src/main/scala/kafka/server/DelayedFetch.scala:
```diff
@@ -71,7 +79,8 @@ class DelayedFetch(delayMs: Long,
+   * Case H: The high watermark on this broker has changed during this request, need to propagate to followers (KIP-392)
    * Upon completion, should return whatever data is available for each valid partition
    */
   override def tryComplete(): Boolean = {
```
Below we have logic which relies on fully realized LogOffsetMetadata (including segment and position) for the high watermark and LSO. Does this work for the follower since it won't have this information in all cases? 

In general, I think our usage of `LogOffsetMetadata` is unsafe since it's unclear when we can expect the segment position to be present. I think it would be better to only expose the `LogOffsetMetadata` object outside of `Log` when we can rely on its full contents.  We could have two methods in Log for this:

```scala
  def highWatermark: Long = highWatermarkMetadata.messageOffset

  def highWatermarkMetadata: LogOffsetMetadata = {
    if (_highWatermarkMetadata.messageOffsetOnly) {
      _highWatermarkMetadata = lookupOffsetMetadata(_highWatermarkMetadata.messageOffset)
    } 
    _highWatermarkMetadata
  }
```

The code can then be modified to only call `highWatermarkMetadata` when we truly need the segment and position.

hachikuji commented on core/src/main/scala/kafka/server/DelayedFetch.scala:
```diff
@@ -114,6 +123,19 @@ class DelayedFetch(delayMs: Long,
             }
+
+            // Case G check HW from follower against latest from leader
+            val followerHW = followerHighwatermarks.apply(topicPartition)
```
In the KIP we proposed that we would return immediately from the first fetch in a session in order to be sure that the follower had an updated high watermark.

also nit: no need to call `apply` directly.

mumrah replied:
Ok, so if we have a request from a follower and no HW information yet (meaning it's the first request in the session), we should just return immediately?

hachikuji replied:
Yeah, does it make sense? Since we're tracking the high watermark in the fetch session, we lose the latest value if we lose the session. So when the session is initialized, we assume the follower doesn't have the latest value. We could probably get around this by tracking the high watermark in the `Replica` object instead. Not sure how easy that would be.

I guess this case should probably be handled in the first fetch in `ReplicaManager.fetchMessages`. So it never actually hits `DelayedFetch`.

hachikuji replied:
Oh, there's one thing we probably have to be careful about. If we are upgrading from an older version, the other broker may not have a fetch session. It would probably be reasonable to skip the high watermark check in this case. Unfortunately, this might mean passing down another flag or something.

hachikuji commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1000,6 +1036,38 @@ class ReplicaManager(val config: KafkaConfig,
+        val replicaEndpoints = metadataCache.getPartitionReplicaEndpoints(tp.topic(), tp.partition(), new ListenerName(clientMetadata.listenerName))
+        val replicaInfoSet: Set[ReplicaView] = partition.allReplicas
+          // Exclude replicas that don't have the requested offset (whether or not if they're in the ISR)
+          .filter(replica => replica.logEndOffset > fetchOffset)
```
Should we check the log start offset as well?

hachikuji replied:
This could probably be `>=`. Mainly we just don't want to send the client to a replica which will just give it an out of range error.

hachikuji reviewed (commented):
Thanks, left a few more comments.

mumrah committed:
Move "leader" to PartitionView, fix DelayedFetch logic

mumrah committed:
Fix DelayedFetch to work on non-leader

mumrah committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-8443

mumrah committed:
Ensure that we get full LogOffsetMetadata when looking at offsets on the follower

mumrah committed:
Fixup after merge from upstream

mumrah committed:
Change last caught up time to time since last caught up

mumrah committed:
More fixup after merge from trunk

mumrah committed:
Don't fast return when HW is missing for non-sessionized requests

mumrah committed:
Fixing a test

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -1512,6 +1547,18 @@ class Log(@volatile var dir: File,
+    * Given a message offset, find its corresponding offset metadata in the log.
+    * If the message offset is out of range, throw an OffsetOutOfRangeException
+    */
+  def convertToOffsetMetadataOrThrow(offset: Long, message: String): LogOffsetMetadata = {
```
Seems it would be more natural to let `convertToOffsetMetadata` call this method instead of the other way around? Otherwise we catch and discard the exception only to re-throw a new one.

mumrah replied:
That's actually what I had originally. I'll switch it back

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -357,12 +357,47 @@ class Log(@volatile var dir: File,
+      val fullOffset = convertToOffsetMetadataOrThrow(_highWatermarkMetadata.messageOffset,
+        "Could not load offset metadata for high watermark")
+      lock.synchronized {
+        _highWatermarkMetadata = fullOffset
```
Hmm.. I think we might need to do the lookup while holding the lock. Otherwise we may overwrite a different value than we are expecting. Same for `lso` below.

Also, we have a related method `maybeFetchHighWatermarkOffsetMetadata`. Perhaps we should try to consolidate?

mumrah replied:
`maybeFetchHighWatermarkOffsetMetadata` is used when we initialized a replica as a leader. It loads the HW metadata and falls back to LSO if it's out of range. Seems like enough of a separate usage to keep separate maybe?

But now that you mention it, should `maybeFetchHighWatermarkOffsetMetadata` be synchronized as well? It probably only gets called by a single thread (Partition#makeLeader), but that doesn't guard against future usages.

hachikuji replied:
I guess it depends on whether the fallback logic makes sense in this context. I guess it is reasonable to raise an out of range error here since we expect the high watermark to exist in the log. 

Agreed that `maybeFetchHighWatermarkOffsetMetadata` should be locked. Possibly we skipped it because this was only used in initialization. I don't see the harm adding it now.

hachikuji replied:
Were we going to add the locking to `initializeHighWatermarkOffsetMetadata`?

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -357,12 +357,47 @@ class Log(@volatile var dir: File,
+    )
+  }
+
+  def offsetSnapshot(full: Boolean = false): LogOffsetSnapshot = {
```
Do we need the `full` option? I think it would be simpler to make the snapshot full always. I looked through the uses and the only case where we use this method but don't need the full snapshot is `ReplicaAlterLogDirsThread.offsetSnapshotFromCurrentReplica`. In this case, it looks like we can just replace this use and access the log start and log end offsets directly.

One other note: `Partition.fetchOffsetSnapshotOrError` is only used by test cases and can probably be removed.

hachikuji commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -917,35 +933,56 @@ class ReplicaManager(val config: KafkaConfig,
-        } else {
-          readInfo.fetchedData
+        // If we are the leader, determine the preferred read-replica
+        val preferredReadReplica = clientMetadata match {
```
nit: you can use `clientMetadata.map`

hachikuji commented on core/src/test/scala/integration/kafka/server/DelayedFetchTest.scala:
```diff
@@ -77,6 +79,81 @@ class DelayedFetchTest extends EasyMockSupport {
     assertEquals(Errors.FENCED_LEADER_EPOCH, fetchResult.error)
   }
 
+  def testCompleteWhenFollowerLaggingHW(followerHW: Option[Long], checkResult: DelayedFetch => Unit): Unit = {
```
Missing `@Test` annotation?

mumrah replied:
This method is actually invoked from a test below it a few times. I'll rename it to avoid confusion

hachikuji replied:
Ah, my bad. Read a bit too quickly.

mumrah committed:
Always get full offset snapshot

more cleanup/feedback

mumrah committed:
Remove unused code

mumrah committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-8443

mumrah commented:
retest this please

hachikuji commented on clients/src/main/java/org/apache/kafka/common/replica/ClientMetadata.java:
```diff
@@ -0,0 +1,95 @@
+    String listenerName();
+
+
+    class DefaultClientMetadata implements ClientMetadata {
```
Is it worthwhile defining `equals` and `hashCode`? Maybe at least a `toString`?

hachikuji commented on clients/src/main/java/org/apache/kafka/common/replica/PartitionView.java:
```diff
@@ -0,0 +1,73 @@
+public interface PartitionView {
+    Set<ReplicaView> replicas();
+
+    Optional<ReplicaView> leader();
```
I'm wondering if the `Optional` is necessary. Would we ever select replicas if we didn't have a current leader?

mumrah replied:
Fair point. I'll update the logic in ReplicaManager and this interface.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/replica/ReplicaView.java:
```diff
@@ -0,0 +1,104 @@
+    long logEndOffset();
+
+    /**
+     * The number of milliseconds (if any) since the last time this replica was caught up to the high watermark.
```
Might be worth mentioning that this is always 0 for the leader?

Also, can we document how this is defined for replicas which have never caught up? Do we use Long.MaxValue or something?

mumrah replied:
If a replica has never caught up, it will return something close to the current time since the caught up time is initialized to zero.

hachikuji commented on clients/src/test/java/org/apache/kafka/common/replica/ReplicaSelectorTest.java:
```diff
@@ -0,0 +1,160 @@
+    }
+
+    static ReplicaView replicaInfo(Node node, long logOffset, long timeSinceLastCaughtUpMs) {
+        return new ReplicaView() {
```
Seems like we could just use `DefaultReplicaView`. Similarly below.

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -357,12 +357,38 @@ class Log(@volatile var dir: File,
+    * offset out of range error if the segment info cannot be loaded.
+    */
   def offsetSnapshot: LogOffsetSnapshot = {
+    val hw: LogOffsetMetadata = lock.synchronized {
```
I think we can avoid the lock in the common case. Maybe we can rewrite this something like the following:
```scala
var hw = this.highWatermark
if (hw.messageOffsetOnly) {
  hw = maybeUpdateHighWatermarkMetadata()
}
```
where `maybeUpdateHighWatermarkMetadata` grabs the lock to safely update the value.

hachikuji commented on core/src/main/scala/kafka/server/DelayedFetch.scala:
```diff
@@ -22,9 +22,10 @@ import java.util.concurrent.TimeUnit
 import org.apache.kafka.common.requests.FetchRequest.PartitionData
 
-import scala.collection._
+import scala.collection.{mutable, _}
```
nit: `mutable` import is not needed

hachikuji commented on core/src/main/scala/kafka/server/DelayedFetch.scala:
```diff
@@ -86,6 +90,7 @@ class DelayedFetch(delayMs: Long,
               expectLeader = fetchMetadata.fetchOnlyLeader)
             val offsetSnapshot = partition.fetchOffsetSnapshot(fetchLeaderEpoch, fetchMetadata.fetchOnlyLeader)
 
+
```
nit: unneeded newline

hachikuji commented on core/src/main/scala/kafka/server/FetchSession.scala:
```diff
@@ -144,6 +144,10 @@ class CachedPartition(val topic: String,
       if (updateResponseData)
         localLogStartOffset = respData.logStartOffset
     }
+    if (respData.preferredReadReplica.isPresent) {
```
If the preferred read replica is the leader, do we need to respond immediately?

mumrah replied:
Is that what this does? I thought `mustRespond` just made sure to include that partition in the response.

hachikuji replied:
I think you're right, but that just changes the question. If the preferred read replica is the leader, do we need to include it in the response? Wouldn't that imply that we always have to include the partition in the response?

mumrah replied:
If the leader picks itself as the preferred read replica, we could exclude it from the response (return none from ReplicaSelector). This actually makes sense b/c we are treating the presence of a preferred read replica ID as a signal to "go somewhere else" to the client and parts of the broker response logic. WDYT?

hachikuji replied:
Yes, that makes sense. So the preferred read replica is only set in the fetch response if it is a node other than the leader.

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -966,6 +968,7 @@ object KafkaConfig {
       .define(InterBrokerSecurityProtocolProp, STRING, Defaults.InterBrokerSecurityProtocol, MEDIUM, InterBrokerSecurityProtocolDoc)
       .define(InterBrokerProtocolVersionProp, STRING, Defaults.InterBrokerProtocolVersion, ApiVersionValidator, MEDIUM, InterBrokerProtocolVersionDoc)
       .define(InterBrokerListenerNameProp, STRING, null, MEDIUM, InterBrokerListenerNameDoc)
+      .define(ReplicaSelectorClassProp, STRING, "", MEDIUM, "")
```
Missing docstring for new config

hachikuji reviewed (commented):
Thanks, left a few more comments.

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -769,7 +791,10 @@ class KafkaApis(val requestChannel: RequestChannel,
-        fetchRequest.isolationLevel)
+        fetchRequest.isolationLevel,
+        clientMetadata,
+        fetchContext.getFollowerHighWatermark,
```
For clients, could we pass through an empty map?

hachikuji commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -875,9 +888,10 @@ class ReplicaManager(val config: KafkaConfig,
-      val delayedFetch = new DelayedFetch(timeout, fetchMetadata, this, quota, responseCallback)
+      val fetchMetadata = FetchMetadata(fetchMinBytes, fetchMaxBytes, hardMaxBytesLimit, isFromFollower,
+        fetchIsolation, isFromFollower, replicaId, fetchPartitionStatus, hasFetchSession)
+      val delayedFetch = new DelayedFetch(timeout, fetchMetadata, this, quota, clientMetadata,
```
If the follower has no cached high watermark, we want to respond immediately. Currently we have this logic in DelayedFetch, but I think it makes more sense to avoid purgatory altogether and respond immediately after the initial fetch above in this function.

mumrah committed:
Finishing touches

mumrah committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-8443

junrao commented on core/src/main/scala/kafka/server/MetadataCache.scala:
```diff
@@ -195,6 +196,22 @@ class MetadataCache(brokerId: Int) extends Logging {
+            case None =>
+              Node.noNode()
+          }}).toMap
+        .filter(pair => !pair._2.isEmpty)
```
Could we use case to avoid unnamed reference _2?

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -85,6 +91,7 @@ case class LogReadResult(info: FetchDataInfo,
                          fetchTimeMs: Long,
                          readSize: Int,
                          lastStableOffset: Option[Long],
+                         preferredReadReplica: Option[Int] = None,
```
Could we add the new param to the javadoc above?

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -818,9 +828,11 @@ class ReplicaManager(val config: KafkaConfig,
                     quota: ReplicaQuota = UnboundedQuota,
                     responseCallback: Seq[(TopicPartition, FetchPartitionData)] => Unit,
-                    isolationLevel: IsolationLevel) {
+                    isolationLevel: IsolationLevel,
```
Could we change the comment above accordingly since we can fetch from any replica now?

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -917,35 +932,54 @@ class ReplicaManager(val config: KafkaConfig,
+          metadata => findPreferredReadReplica(tp, metadata, replicaId, fetchInfo.fetchOffset))
+
+        if (preferredReadReplica.isDefined && !preferredReadReplica.contains(localBrokerId)) {
+          // If the a preferred read-replica is set and is not this replica (the leader), skip the read
```
typo "the a"

junrao commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1000,6 +1034,54 @@ class ReplicaManager(val config: KafkaConfig,
+            .map(leaderNode => new DefaultReplicaView(leaderNode, partition.localLogOrException.logEndOffset, 0L))
+            .get
+
+          replicaInfoSet.add(leaderReplica)
```
The leader replica has been added to replicaInfoSet earlier. Do we need to add it again?

mumrah replied:
It's actually not in there since `Partition#getReplica` returns None for the leader replica after a recent change. I'll update the code to be more clear.

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java:
```diff
@@ -329,7 +329,7 @@ public int hashCode() {
             result = 31 * result + Long.hashCode(lastStableOffset);
             result = 31 * result + Long.hashCode(logStartOffset);
-            result = 31 * result + (preferredReadReplica != null ? preferredReadReplica.hashCode() : 0);
+            result = 31 * result + Objects.hashCode(preferredReadReplica);
```
In FetchRequest.java, it would be useful to add a comment above the following line to summarize the changes.

`    private static final Schema FETCH_REQUEST_V11 = new Schema(
`

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -357,12 +357,38 @@ class Log(@volatile var dir: File,
 
+  /**
+    * Fully materialize and return an offset snapshot including segment position info. This method will update
+    * the LogOffsetMetadata for the high watermark and log start offset if they are message-only. Throws an
```
It seems that we are updating LogOffsetMetadata for last stable offset, not log start offset?

mumrah replied:
Good catch! The "LSO" tripped me up

junrao commented on core/src/main/scala/kafka/server/DelayedFetch.scala:
```diff
@@ -42,7 +43,8 @@ case class FetchMetadata(fetchMinBytes: Int,
                          replicaId: Int,
-                         fetchPartitionStatus: Seq[(TopicPartition, FetchPartitionStatus)]) {
+                         fetchPartitionStatus: Seq[(TopicPartition, FetchPartitionStatus)],
+                         hasFetchSession: Boolean) {
```
Should we add the new param to toString()?

junrao commented on core/src/main/scala/kafka/server/DelayedFetch.scala:
```diff
@@ -114,6 +118,14 @@ class DelayedFetch(delayMs: Long,
+            if (fetchMetadata.isFromFollower && fetchMetadata.hasFetchSession) {
+              // Case G check if the follower has the correct HW from the leader
+              val followerHW = followerHighwatermarks(topicPartition)
+              if (followerHW.isEmpty || followerHW.exists(hw => offsetSnapshot.highWatermark.messageOffset > hw)) {
```
Hmm, it's possible for a follower to always use FullFetchContext due to no available session slots. In that case, followerHW is always empty, but we don't always want to return a fetch response immediately.

junrao reviewed (commented):
@mumrah : Thanks for the PR. Made a pass of the server side changes. A few comments below.

mumrah committed:
Return response immediately if there are no cached HW for the follower

mumrah committed:
Don't return the leader as the preferred read replica

mumrah committed:
Some cleanup from PR feedback

mumrah committed:
Let Replica track the follower's highwatermark rather than FetchSession

This adds highWatermark field to Replica that gets updated along with the other fields
upon a successful read. This is used in ReplicaManager and DelayedFetch to determine
if we should return early due to the follower having an old HW.

hachikuji commented on core/src/main/scala/kafka/cluster/Replica.scala:
```diff
@@ -42,6 +42,10 @@ class Replica(val brokerId: Int, val topicPartition: TopicPartition) extends Log
 
+  // highWatermark is the leader's high watermark after the most recent FetchRequest from this follower. This is
+  // used to determine the maximum HW this follower knows about
+  @volatile private[this] var _highWatermark = 0L
```
Maybe a name like `lastSentHighWatermark` would make it clearer that we do not know this value has actually been acked by the follower. Similar for the getter. This just reduces the risk of making invalid assumptions if we use this in other contexts.

hachikuji commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -869,14 +867,31 @@ class ReplicaManager(val config: KafkaConfig,
+
+    // If the last last sent HW for each requested partitions is behind the HW of the local log,
+    // we should return immediately to ensure the follower has an up-to-date HW
+    val allPartitionsNeedHwUpdate: Boolean = if (isFromFollower) {
```
Have you also considered doing this check in `updateFollowerFetchState`? It is a bit annoying to need another pass over the partitions when we already had the context we needed to make a decision.

mumrah committed:
Simplify the HW check a bit

mumrah committed:
Change the way we update the HW in Replica

mumrah committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-8443

hachikuji commented:
retest this please

hachikuji commented:
retest this please

mumrah committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-8443

mumrah committed:
Add some commentary

mumrah committed:
Increase collection time in test to reduce flakiness

hachikuji commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1521,6 +1527,21 @@ class ReplicaManager(val config: KafkaConfig,
   }
 
+  private def updateFollowerHighWatermark(topicPartition: TopicPartition, followerId: Int, highWatermark: Long): Unit = {
+    nonOfflinePartition(topicPartition) match {
```
We could probably simplify this a little bit. Maybe something like this:
```scala
nonOfflinePartition(topicPartition).map(_.getReplica(followerId)) match {
  case Some(replica) => 
    replica.updateLastSentHighWatermark(highWatermark)
  case None =>
    warn(s"While updating the HW for follower $followerId for partition $topicPartition, " +
         s"the replica could not be found.")
}
```

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -357,12 +357,38 @@ class Log(@volatile var dir: File,
+    * offset out of range error if the segment info cannot be loaded.
+    */
   def offsetSnapshot: LogOffsetSnapshot = {
+    val highWatermark: LogOffsetMetadata = if (_highWatermarkMetadata.messageOffsetOnly) {
```
Technically `_highWatermarkMetadata` might change between the time we call `messageOffsetOnly` and the time we access it below. Maybe we could write it like this:
```scala
var highWatermark = _highWatermarkMetadata
if (highWatermark.messageOffsetOnly) {
  ...
```

hachikuji commented on core/src/main/scala/kafka/server/FetchSession.scala:
```diff
@@ -226,6 +230,10 @@ case class FetchSession(val id: Int,
     Option(partitionMap.find(new CachedPartition(topicPartition))).map(_.fetchOffset)
   }
 
+  def getFollowerHighWatermark(topicPartition: TopicPartition): Option[Long] = synchronized {
```
Seems we didn't need this?

hachikuji commented on core/src/main/scala/kafka/server/MetadataCache.scala:
```diff
@@ -195,6 +196,24 @@ class MetadataCache(brokerId: Int) extends Logging {
+    snapshot.partitionStates.get(topic).flatMap(_.get(partitionId)).map(partitionInfo => {
+      val replicaIds = partitionInfo.basePartitionState.replicas
+      replicaIds.asScala
+        .map(replicaId => replicaId.intValue() -> {
```
We could make this a `flatMap` and skip the `filter` below.

hachikuji commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -856,16 +869,31 @@ class ReplicaManager(val config: KafkaConfig,
     }
 
+    val allPartitionsNeedHwUpdate: Boolean = isFromFollower &&
+      logReadResults.forall {
```
Wouldn't we want to respond if _any_ of the high watermarks need updating?

Also, we can probably turn this into a `var` and check it in the pass above where we are accumulating `bytesReadable`.

hachikuji commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -917,35 +947,60 @@ class ReplicaManager(val config: KafkaConfig,
+
+          // Check if the HW known to the follower is behind the actual HW
+          val followerNeedsHwUpdate: Boolean = partition.getReplica(replicaId)
+            .exists(replica => replica.lastSentHighWatermark < readInfo.highWatermark)
```
Just checking my understanding, but this doesn't account for hw changes as a result of the fetch, right?

mumrah replied:
Correct, since the highWatermark returned in the readInfo is the HW _before_ the read occurs, that's what we're comparing here. If the read caused the HW to advance, we wouldn't know about it here.

hachikuji commented on core/src/main/scala/kafka/server/MetadataCache.scala:
```diff
@@ -195,6 +196,24 @@ class MetadataCache(brokerId: Int) extends Logging {
 
+  def getPartitionReplicaEndpoints(topic: String, partitionId: Int, listenerName: ListenerName): Map[Int, Node] = {
+    val snapshot = metadataSnapshot
+    snapshot.partitionStates.get(topic).flatMap(_.get(partitionId)).map(partitionInfo => {
```
nit: conventionally, we write this as
```scala
.map { partitionInfo =>
...
```

hachikuji commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1000,6 +1055,55 @@ class ReplicaManager(val config: KafkaConfig,
+          val partitionInfo = new DefaultPartitionView(replicaInfoSet.asJava, leaderReplica)
+          replicaSelector.select(tp, clientMetadata, partitionInfo).asScala
+            .filter(!_.endpoint.isEmpty)
+            .filter(!_.equals(leaderReplica))
```
nit: should be able to use `!=`

I think it's worth a comment explaining why we only return the preferred read replica if it is not the leader.

hachikuji commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1000,6 +1055,55 @@ class ReplicaManager(val config: KafkaConfig,
+        Option.empty
+      } else {
+        val replicaEndpoints = metadataCache.getPartitionReplicaEndpoints(tp.topic(), tp.partition(), new ListenerName(clientMetadata.listenerName))
+        val now = time.milliseconds
```
nit: we can pass `fetchTimeMs` from `readFromLocalLog`

mumrah committed:
More PR feedback

mumrah committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-8443

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -357,12 +357,38 @@ class Log(@volatile var dir: File,
+    var highWatermark = _highWatermarkMetadata
+    if (highWatermark.messageOffsetOnly) {
+      lock.synchronized {
+        val fullOffset = convertToOffsetMetadataOrThrow(highWatermark.messageOffset)
```
This is pretty tough to get right. I think we have to use `_highWatermarkMetadata` here since the value may change before acquiring the lock.

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -357,12 +357,38 @@ class Log(@volatile var dir: File,
+      }
+    }
+
+    val lastStable: LogOffsetMetadata = if (firstUnstableOffset.exists(_.messageOffsetOnly)) {
```
I think we have the same "check then act" problem here. This is also a bit tricky because we are depending indirectly on the high watermark if the first unstable offset is not defined. I _think_ this works:

```scala
var lso = lastStableOffsetMetadata
if (lso.messageOffsetOnly) {
  lock synchronized {
    firstUnstableOffset match {
      case None => highWatermark
      case Some(offsetMetadata) =>
        val fullOffset = convertToOffsetMetadataOrThrow(offsetMetadata.messageOffset)
        firstUnstableOffset = Some(fullOffset)
        lso = fullOffset
    }
  }
}
```

hachikuji commented on core/src/main/scala/kafka/server/MetadataCache.scala:
```diff
@@ -211,7 +211,7 @@ class MetadataCache(brokerId: Int) extends Logging {
           case (_, node) => !node.isEmpty
         })
-    }).getOrElse(Map.empty[Int, Node])
+    } getOrElse Map.empty[Int, Node]
```
nit: this syntax tends to be only used for infix operators like `+`

mumrah committed:
Fix locking for HW and LSO in Log

mumrah committed:
More cleanup

mumrah committed:
Remove LeaderReplicaSelector, make that the default behavior

mumrah committed:
Add logging

hachikuji reviewed (approved):
LGTM. Thanks for the patch!

mumrah committed:
Separating out the system test

hachikuji commented:
retest this please

hachikuji commented:
Submitted #7033 for previous failures.
